### PR TITLE
Run job and job run status

### PR DESCRIPTION
**Pull request by agaffney**
* support multiple steps in a test
* support preserving keys from response in one step to use in another
* support for specifying HTTP status code in JsonError()
* return 404 in cases where we are given a specific job ID and get an error

### CODE REVIEW

File: src/helpers/helpers.go [hunk 1]
```diff
@@ -69,6 +79,26 @@ func JobKubernetesToMetronome(job *v1beta1.CronJob) *MetronomeJob {
 	return ret
 }
 
+// Convert Kuberentes Job to Metronome job run format
```
Comment (wolf31o2) [inline src/helpers/helpers.go]: Typo

Comment (agaffney) [reply src/helpers/helpers.go]: Fixed

File: src/api/runs.go [hunk 2]
```diff
@@ -9,11 +12,39 @@ func handleGetJobRuns(c *gin.Context) {
...
+	if err != nil {
+		JsonError(c, fmt.Sprintf("cannot run job: %s", err))
+		return
+	}
+	c.JSON(201, helpers.JobKubernetesToMetronome(job))
 }
 
 func handleGetJobRun(c *gin.Context) {
-	c.String(200, "TODO")
+	//jobId := c.Param("jobid")
```
Comment (tmad) [inline src/api/runs.go]: delete this

File: src/helpers/helpers.go [hunk 3]
```diff
@@ -69,6 +79,26 @@ func JobKubernetesToMetronome(job *v1beta1.CronJob) *MetronomeJob {
...
+func JobKubernetesToMetronome(job *batchv1.Job) *MetronomeJobRun {
+	ret := &MetronomeJobRun{}
+	ret.Id = fmt.Sprintf("%s.%s", job.ObjectMeta.Namespace, job.ObjectMeta.Name)
+	ret.JobId = fmt.Sprintf("%s.%s", job.ObjectMeta.Namespace, job.ObjectMeta.OwnerReferences[0].Name)
+	ret.CreatedAt = job.ObjectMeta.CreationTimestamp.String()
+	if job.Status.StartTime == nil {
+		ret.Status = "STARTING"
+	} else if job.Status.CompletionTime == nil {
+		ret.Status = "RUNNING"
+	} else {
```
Comment (tmad) [inline src/helpers/helpers.go]: optional: look for job.Status.Failed and set ret.Status to FAILED

File: src/api/runs.go [hunk 4]
```diff
@@ -9,11 +12,39 @@ func handleGetJobRuns(c *gin.Context) {
...
+	runId := c.Param("runid")
+	namespace, name, err := helpers.SplitMetronomeJobId(runId)
+	if err != nil {
+		JsonError(c, err.Error())
+		return
+	}
+	job, err := kube.GetJob(namespace, name)
+	if err != nil {
+		JsonError(c, fmt.Sprintf("cannot get job: %s", err))
```
Comment (tmad) [inline src/api/runs.go]: metronome api specifies to throw 404 here (unless you expect it to fail due to different reasons than job not existing)

Comment (agaffney) [reply src/api/runs.go]: The kube library just returns an error and doesn't really differentiate between "not found" and error, but you're probably right about it being more likely that an error here just means the job isn't found. However, this logic would need to be fixed in more than just this one place.

File: src/helpers/helpers.go (hunk 3 above)
Comment (agaffney) [reply src/helpers/helpers.go]: It seems that `job.Status.Failed` is an `int32` vs. a `bool`, so I'm not entirely sure how it's supposed to work.

**Pull request merged**